### PR TITLE
Fix regression in StreamWriter.Write perf for small inputs

### PR DESCRIPTION
**Pull request by stephentoub**
My previous changes to add span support to StreamWriter regressed performance for small inputs.  This was primarily due to refactoring out a bunch of duplicated code into a shared span-based helper.

This change keeps the helper but uses AggressiveInlining to avoid the extra function call, essentially restoring the previous "every method has its own copy" approach but without the C# code duplication.  I also cleaned up the CheckAsyncTaskInProgress helper to make it streamlined and inlineable.

| Test                                                                      | Before      | After    | Improvement | 
|---------------------------------------------------------------------------|-------------|----------|-------------| 
| System.IO.Tests.Perf_StreamWriter.WriteCharArray(writeLength: 100)        | 1041.38 | 971.10  | 6.75%       | 
| System.IO.Tests.Perf_StreamWriter.WriteCharArray(writeLength: 2)          | 168.70 | 109.17  | 35.29%      | 
| System.IO.Tests.Perf_StreamWriter.WritePartialCharArray(writeLength: 100) | 1096.65 | 1001.64 | 8.66%       | 
| System.IO.Tests.Perf_StreamWriter.WritePartialCharArray(writeLength: 2)   | 225.62 | 153.26  | 32.07%      | 
| System.IO.Tests.Perf_StreamWriter.WriteString(writeLength: 100)           | 1095.40 | 987.03  | 9.89%       | 
| System.IO.Tests.Perf_StreamWriter.WriteString(writeLength: 2)             | 192.74 | 114.06  | 40.82%      | 

Fixes https://github.com/dotnet/corefx/issues/27496
cc: @danmosemsft, @jkotas

### CODE REVIEW

File: src/mscorlib/shared/System/IO/StreamWriter.cs [hunk 1]
```diff
@@ -325,10 +328,7 @@ public override void Write(char value)
 
         public override void Write(char[] buffer)
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/StreamWriter.cs]: Is the inliner smart enough to not inline WriteSpan when it inlines (de-virtualized) Write?

Comment (stephentoub) [reply src/mscorlib/shared/System/IO/StreamWriter.cs]: It appears it inlines WriteSpan into Write and then doesn't inline Write into its caller, as desired.  Do you want me to add NoInlining to all of the call sites of WriteSpan?
cc: @AndyAyersMS

**Comment by danmosemsft** [comment]
This also more than reverses the more recent 11% regression in WriteString(writeLength: 2) - did not determine the cause of that. Good!

File: src/mscorlib/shared/System/IO/StreamWriter.cs (hunk 1 above)
Comment (AndyAyersMS) [reply src/mscorlib/shared/System/IO/StreamWriter.cs]: I would expect if `Write` gets devirtualized and inlined then `WriteSpan` would also get inlined, as `AggressiveInlining` doesn't get cancelled out. But it's probably not super-common for `Write` to be devirtualized as we have to know we have exactly a `StreamReader` (say, the constructor call needs to be in the same method).

Adding `NoInlining` to the callers of `WriteSpan` is a reasonable way to limit the amount of code that gets duplicated.

Comment (stephentoub) [reply src/mscorlib/shared/System/IO/StreamWriter.cs]: Thanks. Added.

**Pull request merged**